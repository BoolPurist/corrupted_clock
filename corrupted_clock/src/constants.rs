#[macro_export]
macro_rules! env_var_prefix {
    ($suffix:literal) => {{
        concat!("CORRUPTED_CLOCK_", $suffix)
    }};
}

pub use env_var_prefix;

pub const STOP_WATCH_ALIASE: &str = "sw";
pub const COUNT_DOWN_ALIASE: &str = "cd";
pub const ALL_CLOCK_ALIASE: &str = "a";
