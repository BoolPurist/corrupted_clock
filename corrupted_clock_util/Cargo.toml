[package]
name = "corrupted_clock_util"
version = "0.2.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
thiserror.workspace = true
serde_json.workspace = true
 
chrono = { version = "0.4.38", features = ["serde"] } 
serde = { version = "1.0.199", features = ["derive"] } 
regex = "1.10.4"
once_cell = "1.19.0"

[dev-dependencies]
insta = { version = "1.38.0", features = ["json"] }
pretty_assertions = "1.4.0"
