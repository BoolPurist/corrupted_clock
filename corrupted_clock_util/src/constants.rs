pub const APP_NAME: &str = "corrupted_clock";
pub const STATE_FILE_NAME: &str = "app_state.json";
