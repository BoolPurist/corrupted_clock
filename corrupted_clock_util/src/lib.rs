use chrono::{DateTime, Datelike, Local, Timelike, Utc};

pub mod constants;
pub mod data_store;
pub mod parsed_date;
pub mod prelude;
pub mod timing;

pub fn convert_utc_to_local(date_time: DateTime<Utc>) -> DateTime<Local> {
    date_time.into()
}

pub fn local_now() -> DateTime<Local> {
    Local::now()
}

pub fn chrono_time_to_str<T>(date_time: DateTime<T>) -> String
where
    T: chrono::TimeZone,
{
    format!(
        "{:04}-{:02}-{:02} {:02}:{:02}:{:02}",
        date_time.year(),
        date_time.month(),
        date_time.day(),
        date_time.hour(),
        date_time.minute(),
        date_time.second(),
    )
}
