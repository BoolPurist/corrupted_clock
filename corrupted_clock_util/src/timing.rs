pub use clock_duration::ClockDuration;
pub use clock_table::ClockTable;
pub use count_down::CountDown;
use serde::{Deserialize, Serialize};
pub use stopwatch::Stopwatch;

pub mod mocking_time;
#[cfg(test)]
pub mod test_utils;

mod clock_duration;
mod clock_table;
mod count_down;
mod stopwatch;

pub type UtcDateTime = DateTime<Utc>;
pub type ChronoDuration = chrono::Duration;

use chrono::{DateTime, Utc};
use thiserror::Error;

#[derive(Default, Serialize, Deserialize)]
pub struct UtcTimeImpl;

#[derive(Debug, Error, Clone, Copy)]
#[error("'{0}' as a start date must not be in the future")]
pub struct InvalidDateInFuture(UtcDateTime);

pub fn validate_if_date_is_not_in_future(
    time: &impl TimeImpl,
    to_check: UtcDateTime,
) -> Result<(), InvalidDateInFuture> {
    let now = time.now();
    if now.timestamp() < to_check.timestamp() {
        Err(InvalidDateInFuture(to_check))
    } else {
        Ok(())
    }
}

pub trait Timer {
    fn created_at(&self) -> UtcDateTime;
    fn start_moment(&self) -> UtcDateTime;
    fn last_resumed_at(&self) -> Option<UtcDateTime>;
    fn last_paused_at(&self) -> Option<UtcDateTime>;
    fn passed(&self) -> ClockDuration;
    fn paused_time(&self) -> ClockDuration;
    fn is_paused(&self) -> bool;
    fn pause(&mut self);
    fn resume(&mut self);
    fn reset(&mut self);
}

pub trait TimeImpl {
    fn now(&self) -> UtcDateTime;
}

impl<'a, T> TimeImpl for &'a T
where
    T: TimeImpl,
{
    fn now(&self) -> UtcDateTime {
        (*self).now()
    }
}

impl TimeImpl for UtcTimeImpl {
    fn now(&self) -> UtcDateTime {
        Utc::now()
    }
}
