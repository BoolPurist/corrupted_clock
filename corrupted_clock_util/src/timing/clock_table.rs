use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use thiserror::Error;

use super::{CountDown, Stopwatch};

#[derive(Serialize, Deserialize, Default, Debug)]
pub struct ClockTable {
    stopwatches: HashMap<String, Stopwatch>,
    count_downs: HashMap<String, CountDown>,
}

#[derive(Debug, Error)]
#[error("Key {0} is already taken")]
pub struct AlreadyAdded(String);

impl ClockTable {
    pub fn all_stopwatches(&self) -> impl Iterator<Item = (&str, &Stopwatch)> {
        self.stopwatches
            .iter()
            .map(|(key, value)| (key.as_str(), value))
    }

    pub fn all_count_downs(&self) -> impl Iterator<Item = (&str, &CountDown)> {
        self.count_downs
            .iter()
            .map(|(key, value)| (key.as_str(), value))
    }

    pub fn remove_count_down(&mut self, key: &str) -> bool {
        self.count_downs.remove(key).is_some()
    }

    pub fn has_stop_watch(&self, key: &str) -> bool {
        self.stopwatches.contains_key(key)
    }

    pub fn has_count_down(&self, key: &str) -> bool {
        self.count_downs.contains_key(key)
    }

    pub fn remove_stopwatch(&mut self, key: &str) -> bool {
        self.stopwatches.remove(key).is_some()
    }

    pub fn modify_stopwatch(&mut self, key: &str) -> Option<&mut Stopwatch> {
        self.stopwatches.get_mut(key)
    }

    pub fn modify_count_down(&mut self, key: &str) -> Option<&mut CountDown> {
        self.count_downs.get_mut(key)
    }

    pub fn get_stopwatch(&self, key: &str) -> Option<&Stopwatch> {
        self.stopwatches.get(key)
    }

    pub fn get_count_down(&self, key: &str) -> Option<&CountDown> {
        self.count_downs.get(key)
    }

    pub fn add_stopwatch(&mut self, key: String, sw: Stopwatch) -> Result<(), AlreadyAdded> {
        if self.stopwatches.contains_key(&key) {
            return Err(AlreadyAdded(key));
        }
        _ = self.stopwatches.insert(key, sw);
        Ok(())
    }

    pub fn add_count_down(&mut self, key: String, sw: CountDown) -> Result<(), AlreadyAdded> {
        if self.count_downs.contains_key(&key) {
            return Err(AlreadyAdded(key));
        }
        _ = self.count_downs.insert(key, sw);
        Ok(())
    }
}
