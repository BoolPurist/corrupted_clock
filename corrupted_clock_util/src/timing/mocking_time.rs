use std::{cell::RefCell, rc::Rc};

use chrono::TimeDelta;

use super::{TimeImpl, UtcDateTime};

#[derive(Default, Clone, Debug)]
pub struct MockTimeImpl(Rc<RefCell<UtcDateTime>>);

impl TimeImpl for MockTimeImpl {
    fn now(&self) -> UtcDateTime {
        self.0.borrow().clone()
    }
}

impl MockTimeImpl {
    pub fn new(current_now: UtcDateTime) -> Self {
        let now = Rc::new(RefCell::new(current_now));
        Self(now)
    }

    pub fn add_to_now(&self, to_add: TimeDelta) {
        *self.0.borrow_mut() += to_add;
    }
    pub fn set_now(&self, new_now: UtcDateTime) {
        *self.0.borrow_mut() = new_now;
    }
}
