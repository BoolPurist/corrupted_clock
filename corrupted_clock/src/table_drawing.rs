use corrupted_clock_util::timing::{CountDown, Stopwatch, TimeImpl, Timer, UtcDateTime};
use prettytable::{Cell, Row, Table};

use crate::constants::NOT_AVIABLE_TXT;

pub fn count_down_rows<'a, T>(
    count_downs: impl IntoIterator<Item = (&'a str, &'a CountDown<T>)>,
) -> Table
where
    T: Default + TimeImpl + 'a,
{
    let count_downs = count_downs.into_iter();
    let mut table = Table::new();
    let header = stopwatch_header().chain(count_down_headers()).collect();
    table.add_row(Row::new(header));
    for name_count_down in count_downs {
        let (_, cd) = name_count_down;
        let row = stopwatch_fields(name_count_down).chain(count_down_fields(cd));
        table.add_row(row.collect());
    }
    table
}

pub fn stop_watch_rows<'a, T>(
    count_downs: impl IntoIterator<Item = (&'a str, &'a Stopwatch<T>)>,
) -> Table
where
    T: Default + TimeImpl + 'a,
{
    type CellVec = Vec<Cell>;
    let stop_watches = count_downs.into_iter();

    let mut table = Table::new();
    let headers: CellVec = stopwatch_header().collect();
    let elements_num = headers.len();
    table.add_row(Row::new(headers));

    for name_stop_watch in stop_watches {
        let fields: CellVec = stopwatch_fields(name_stop_watch)
            .map(|e| Cell::new(&e))
            .collect();
        debug_assert_eq!(elements_num, fields.len());
        table.add_row(fields.into());
    }
    table
}

fn stopwatch_header() -> impl Iterator<Item = Cell> {
    [
        "Name",
        "Created at",
        "Started at",
        "Is paused",
        "Passed Time",
        "Paused Time",
        "Last resumed at",
        "Last paused at",
    ]
    .map(Cell::new)
    .into_iter()
}

fn stopwatch_fields<'a, T>((name, stop_watch): (&'a str, &'a T)) -> impl Iterator<Item = String>
where
    T: Timer + 'a,
{
    fn to_local_short_table_field(date: UtcDateTime) -> String {
        corrupted_clock_util::chrono_time_to_str(corrupted_clock_util::convert_utc_to_local(date))
    }

    fn convert_to_opt_table_field(opt: Option<UtcDateTime>) -> String {
        if let Some(date) = opt {
            to_local_short_table_field(date)
        } else {
            NOT_AVIABLE_TXT.to_string()
        }
    }

    let local_create_at = to_local_short_table_field(stop_watch.created_at());
    let local_started_at = to_local_short_table_field(stop_watch.start_moment());
    let last_resumed = convert_to_opt_table_field(stop_watch.last_resumed_at());
    let last_paused = convert_to_opt_table_field(stop_watch.last_paused_at());

    [
        name.to_string(),
        local_create_at,
        local_started_at,
        stop_watch.is_paused().to_string(),
        stop_watch.passed().to_string(),
        stop_watch.paused_time().to_string(),
        last_resumed,
        last_paused,
    ]
    .into_iter()
}

fn count_down_headers() -> impl Iterator<Item = Cell> {
    ["Count down", "Left Time"].map(Cell::new).into_iter()
}

fn count_down_fields<'a, T>(cd: &CountDown<T>) -> impl Iterator<Item = String>
where
    T: Default + TimeImpl,
{
    [cd.count_down_time().to_string(), cd.left_time().to_string()].into_iter()
}
