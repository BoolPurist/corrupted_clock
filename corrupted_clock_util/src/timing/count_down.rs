use serde::{Deserialize, Serialize};

use super::{
    ClockDuration, InvalidDateInFuture, Stopwatch, TimeImpl, Timer, UtcDateTime, UtcTimeImpl,
};

#[derive(Serialize, Deserialize)]
pub struct CountDown<T = UtcTimeImpl>
where
    T: Default,
{
    stopwatch: Stopwatch<T>,
    time: ClockDuration,
}

impl<T> std::fmt::Debug for CountDown<T>
where
    T: Default,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("CountDown")
            .field("stopwatch", &self.stopwatch)
            .field("time", &self.time)
            .finish()
    }
}

impl CountDown {
    pub fn new(time: ClockDuration) -> Self {
        let stopwatch: Stopwatch = Stopwatch::new();
        Self { stopwatch, time }
    }

    pub fn new_with_start(
        time: ClockDuration,
        start: UtcDateTime,
    ) -> Result<Self, InvalidDateInFuture> {
        let stopwatch: Stopwatch = Stopwatch::new_with_impl_and_start_date(UtcTimeImpl, start)?;
        Ok(Self { stopwatch, time })
    }
}

impl<T> CountDown<T>
where
    T: TimeImpl + Default,
{
    pub fn count_down_time(&self) -> ClockDuration {
        self.time
    }

    pub fn new_with_impl(time_impl: T, time: ClockDuration) -> Self {
        let stopwatch = Stopwatch::new_with_impl(time_impl);
        Self { stopwatch, time }
    }

    pub fn left_time(&self) -> ClockDuration {
        let left = self.time - self.passed();
        debug_assert!(left.total_secs() >= 0);
        debug_assert!(left.nanos_part() >= 0);
        left
    }

    pub fn stopwatch(&self) -> &Stopwatch<T> {
        &self.stopwatch
    }
}

impl<T> Timer for CountDown<T>
where
    T: TimeImpl + Default,
{
    fn passed(&self) -> ClockDuration {
        let actual_passed = self.stopwatch.passed();
        actual_passed.min(self.time)
    }

    fn pause(&mut self) {
        self.stopwatch.pause();
    }

    fn resume(&mut self) {
        self.stopwatch.resume();
    }

    fn reset(&mut self) {
        self.stopwatch.reset();
    }

    fn is_paused(&self) -> bool {
        self.stopwatch.is_paused()
    }

    fn created_at(&self) -> super::UtcDateTime {
        self.stopwatch().created_at()
    }

    fn paused_time(&self) -> ClockDuration {
        self.stopwatch().paused_time()
    }

    fn last_resumed_at(&self) -> Option<super::UtcDateTime> {
        self.stopwatch.last_resumed_at()
    }

    fn last_paused_at(&self) -> Option<super::UtcDateTime> {
        self.stopwatch.last_paused_at()
    }

    fn start_moment(&self) -> super::UtcDateTime {
        self.stopwatch.start_moment()
    }
}

#[cfg(test)]
mod testing {
    use chrono::TimeDelta;

    use crate::timing::{
        mocking_time::MockTimeImpl,
        test_utils::{add_to_now, new_utc_moment},
    };

    use super::*;

    fn set_up_counte_mock(
        input: &str,
        duration: ClockDuration,
    ) -> (CountDown<MockTimeImpl>, MockTimeImpl) {
        let time_impl = MockTimeImpl::new(new_utc_moment(input));
        (
            CountDown::new_with_impl(time_impl.clone(), duration),
            time_impl,
        )
    }

    fn assert_left_time(count_down: &CountDown<MockTimeImpl>, expected: ClockDuration) {
        let left_time = count_down.left_time();
        pretty_assertions::assert_eq!(expected, left_time);
    }

    #[test]
    fn count_down_to_left_time() {
        let intial_time = ClockDuration::new_secs_mins_hours(Some(2), None, None);
        let (count_down, mut setter) = set_up_counte_mock("2020-02-11 12:00:00", intial_time);
        assert_left_time(&count_down, intial_time);
        add_to_now(&mut setter, TimeDelta::hours(1));
        add_to_now(&mut setter, TimeDelta::hours(2));
        // Count should be finished now
        assert_left_time(&count_down, ClockDuration::default());
    }
}
