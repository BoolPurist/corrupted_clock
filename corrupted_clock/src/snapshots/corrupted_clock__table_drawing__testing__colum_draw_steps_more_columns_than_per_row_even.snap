---
source: corrupted_clock/src/table_drawing.rs
expression: actual
---
[
    4,
    4,
]
